### PR TITLE
Java codegen daml lf map

### PR DESCRIPTION
**Pull request by nicholassmith-da**
Generate Java code for DAML-LF Map primitives.

Wip as implementation of DAML primitive is still under development.

### CODE REVIEW

File: language-support/java/codegen/src/it/java-1.3/com.digitalasset/AllTests.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset;
+
+import org.junit.runner.RunWith;
+import org.junit.runners.Suite;
+
+@RunWith(Suite.class)
+@Suite.SuiteClasses({
+        OptionalTest.class,
```
Comment (gerolf-da) [inline language-support/java/codegen/src/it/java-1.3/com.digitalasset/AllTests.java]: We don't need `OptionalTest` for 1.3, as nothing has changed in that regard.
`MapTest.class` needs to be added.

File: language-support/java/codegen/src/it/java-1.3/com.digitalasset/OptionalTest.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+import tests.optionaltest.optionalvariant.OptionalPrimVariant;
+
+import java.util.Arrays;
+import java.util.Optional;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+
+@RunWith(JUnitPlatform.class)
+public class OptionalTest {
```
Comment (gerolf-da) [inline language-support/java/codegen/src/it/java-1.3/com.digitalasset/OptionalTest.java]: `OptionalTest` is not needed for 1.3.

File: language-support/java/codegen/src/it/java/com/digitalasset/ListTest.java [hunk 3]
```diff
@@ -146,4 +149,78 @@ void listOfColorsFromProtobufValue() {
 
     }
 
+    @Test
+    void parameterizedListTestRoundTrip() {
```
Comment (gerolf-da) [inline language-support/java/codegen/src/it/java/com/digitalasset/ListTest.java]: Can you move this test to a different PR independent from the introduction of Map?

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/Main.scala [hunk 4]
```diff
@@ -13,7 +13,7 @@ object StandaloneMain {
     try {
       Main.main(args)
     } catch {
-      case NonFatal(_) => sys.exit(-1)
+      case NonFatal(x) => { x.printStackTrace(); sys.exit(-2) }
```
Comment (gerolf-da) [inline language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/Main.scala]: We probably don't want to keep that.

File: language-support/java/codegen/src/it/java/com/digitalasset/ListTest.java (hunk 3 above)
Comment (nicholassmith-da) [reply language-support/java/codegen/src/it/java/com/digitalasset/ListTest.java]: Good idea.

File: language-support/java/codegen/src/it/java-1.3/com.digitalasset/AllTests.java (hunk 1 above)
Comment (nicholassmith-da) [reply language-support/java/codegen/src/it/java-1.3/com.digitalasset/AllTests.java]: Personally I would keep the `OptionalTest` to test for regressions (although upstream modules should be also be doing this). The copy-n-paste isn't ideal and the structuring of the DAML-LF version specific tests could (should) be improved.

Comment (gerolf-da) [reply language-support/java/codegen/src/it/java-1.3/com.digitalasset/AllTests.java]: Agreed. But let's fix the test setup in another PR without copy-pasta. wdyt?

Comment (nicholassmith-da) [reply language-support/java/codegen/src/it/java-1.3/com.digitalasset/AllTests.java]: Also good.

File: docs/source/app-dev/bindings-java/codegen.rst [hunk 5]
```diff
@@ -109,7 +109,8 @@ Java:
 +-----------------------------------+---------------------------------------+
 | ``List`` or ``[]``                | ``java.util.List``                    |
 +-----------------------------------+---------------------------------------+
-| ``TextMap``                       | ``java.util.Map`` (with ``String Keys``) |
+| ``TextMap``                       | ``java.util.Map`` i                   |
```
Comment (gerolf-da) [inline docs/source/app-dev/bindings-java/codegen.rst]: There's a stray `i`.

File: docs/source/app-dev/bindings-java/codegen.rst [hunk 6]
```diff
@@ -109,6 +109,9 @@ Java:
 +-----------------------------------+---------------------------------------+
 | ``List`` or ``[]``                | ``java.util.List``                    |
 +-----------------------------------+---------------------------------------+
+| ``TextMap``                       | ``java.util.Map`` i                   |
+|                                   | Restricted to using ``String'' keys`` |
```
Comment (gerolf-da) [inline docs/source/app-dev/bindings-java/codegen.rst]:
```suggestion
|                                   | Restricted to using ``String`` keys   |
```

File: language-support/java/codegen/src/it/java-1.3/com.digitalasset/MapTest.java [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+        ValueOuterClass.Record protoRecord = ValueOuterClass.Record.newBuilder()
+                .addFields(ValueOuterClass.RecordField.newBuilder()
+                        .setLabel("mapField")
+                        .setValue(ValueOuterClass.Value.newBuilder()
+                                .setMap(ValueOuterClass.Map.newBuilder()
+                                        .addEntries(ValueOuterClass.Map.Entry.newBuilder()
+                                                .setKey(ValueOuterClass.Value.newBuilder()
+                                                        .setText("outerkey1"))
+                                                .setValue(ValueOuterClass.Value.newBuilder()
+                                                        .setRecord(ValueOuterClass.Record.newBuilder()
```
Comment (gerolf-da) [inline language-support/java/codegen/src/it/java-1.3/com.digitalasset/MapTest.java]: The value of the outer map should be another `Map` and not a `Record`.

File: language-support/java/codegen/src/it/java-1.3/com.digitalasset/MapTest.java [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+                                                                        .build())
+                                                                .build())
+                                                        .build())
+                                                .build()
+                                        )
+                                        .addEntries(ValueOuterClass.Map.Entry.newBuilder()
+                                                .setKey(ValueOuterClass.Value.newBuilder()
+                                                        .setText("outerkey1"))
+                                                .setValue(ValueOuterClass.Value.newBuilder()
+                                                        .setRecord(ValueOuterClass.Record.newBuilder()
```
Comment (gerolf-da) [inline language-support/java/codegen/src/it/java-1.3/com.digitalasset/MapTest.java]: Same as above, the value of the outer map should be another `Map` and not a `Record`.

File: language-support/java/codegen/src/it/java-1.3/com.digitalasset/MapTest.java [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+                                .addEntries(ValueOuterClass.Map.Entry.newBuilder()
+                                        .setKey(ValueOuterClass.Value.newBuilder().setText("key"))
+                                        .setValue(ValueOuterClass.Value.newBuilder().setInt64(42L))
+                                        .build())
+                                .build())
+                        .build())
+                .build();
+
+        Variant dataVariant = Variant.fromProto(protoVariant);
+        ParameterizedVariant<?> fromValue = ParameterizedVariant.fromValue(dataVariant,
```
Comment (gerolf-da) [inline language-support/java/codegen/src/it/java-1.3/com.digitalasset/MapTest.java]: Replace `?` with `Long`

File: language-support/java/codegen/src/it/java-1.3/com.digitalasset/MapTest.java [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+                f -> f.asInt64().orElseThrow(() -> new IllegalArgumentException("Expected Long value")));
```
Comment (gerolf-da) [inline language-support/java/codegen/src/it/java-1.3/com.digitalasset/MapTest.java]: To make it work with `Long` we need to add a `getValue()` after `orElseThrow()`.

File: language-support/java/codegen/src/it/java-1.3/com.digitalasset/MapTest.java [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+        ParameterizedVariant<Long> fromConstructor = new ParameterizedVariant<>(Collections.singletonMap("key", 42L));
+
+        assertEquals(fromValue, fromConstructor);
+        assertEquals(fromConstructor.toValue(f -> new Int64(f)), dataVariant);
```
Comment (gerolf-da) [inline language-support/java/codegen/src/it/java-1.3/com.digitalasset/MapTest.java]: Add the same assertion with `fromValue` instead of `fromConstructor`.

File: language-support/java/codegen/src/it/java-1.3/com.digitalasset/MapTest.java [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset;
+
+import com.daml.ledger.javaapi.data.Int64;
+import com.daml.ledger.javaapi.data.Record;
+import com.daml.ledger.javaapi.data.Text;
+import com.daml.ledger.javaapi.data.Variant;
+import com.digitalasset.ledger.api.v1.ValueOuterClass;
+import org.junit.Test;
```
Comment (gerolf-da) [inline language-support/java/codegen/src/it/java-1.3/com.digitalasset/MapTest.java]:
```suggestion
import org.junit.jupiter.api.Test;
```
